extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+  {
+    this.selfDruidNode = context.selfNode();
+    context.registerController(this, closer);
+
+    this.netClient = new ExceptionWrappingWorkerClient(context.taskClientFor(this));
+    closer.register(netClient::close);
+
+    final boolean isDurableStorageEnabled =
+        MultiStageQueryContext.isDurableStorageEnabled(task.getQuerySpec().getQuery().getContext());
+
+    final QueryDefinition queryDef = makeQueryDefinition(
+        id(),
+        makeQueryControllerToolKit(),
+        task.getQuerySpec()
+    );
+
+    QueryValidator.validateQueryDef(queryDef);
+    queryDefRef.set(queryDef);
+
+    log.debug("Query [%s] durable storage mode is set to %s.", queryDef.getQueryId(), isDurableStorageEnabled);
+
+    this.workerTaskLauncher = new MSQWorkerTaskLauncher(
+        id(),
+        task.getDataSource(),
+        context,
+        isDurableStorageEnabled,
+
+        // 10 minutes +- 2 minutes jitter
+        TimeUnit.SECONDS.toMillis(600 + ThreadLocalRandom.current().nextInt(-4, 5) * 30L)
+    );
+
+    long maxParseExceptions = -1;
+
+    if (task.getSqlQueryContext() != null) {
+      maxParseExceptions = Optional.ofNullable(
+                                       task.getSqlQueryContext().get(MSQWarnings.CTX_MAX_PARSE_EXCEPTIONS_ALLOWED))
+                                   .map(DimensionHandlerUtils::convertObjectToLong)
+                                   .orElse(MSQWarnings.DEFAULT_MAX_PARSE_EXCEPTIONS_ALLOWED);
+    }
+
+    this.faultsExceededChecker = new FaultsExceededChecker(
+        ImmutableMap.of(CannotParseExternalDataFault.CODE, maxParseExceptions)
+    );
+
+    return queryDef;
+  }
+
+  private Pair<ControllerQueryKernel, ListenableFuture<?>> runQueryUntilDone(
+      final QueryDefinition queryDef,
+      final InputSpecSlicerFactory inputSpecSlicerFactory,
+      final Closer closer
+  ) throws Exception
+  {
+    // Start tasks.
+    log.debug("Query [%s] starting tasks.", queryDef.getQueryId());
+
+    final ListenableFuture<?> workerTaskLauncherFuture = workerTaskLauncher.start();
+    closer.register(() -> workerTaskLauncher.stop(true));
+
+    workerTaskLauncherFuture.addListener(
+        () ->
+            addToKernelManipulationQueue(queryKernel -> {
+              // Throw an exception in the main loop, if anything went wrong.
+              FutureUtils.getUncheckedImmediately(workerTaskLauncherFuture);
+            }),
+        Execs.directExecutor()
+    );
+
+    // Segments to generate; used for making stage-two workers.
+    List<SegmentIdWithShardSpec> segmentsToGenerate = null;
+
+    // Track which stages have got their partition boundaries sent out yet.
+    final Set<StageId> stageResultPartitionBoundariesSent = new HashSet<>();
+
+    // Start query tracking loop.
+    log.debug("Query [%s] starting tracker.", queryDef.getQueryId());
+    final ControllerQueryKernel queryKernel = new ControllerQueryKernel(queryDef);
+
+    while (!queryKernel.isDone()) {
```

Comment:
An explanation here would help. I seems this loop iterates per stage? For each stage, it starts a stage, tells the kernel about the stages, assigns shuffles, loops to monitor and report on each worker, then issues reports if the query is done, else loops for the next stage. Or something like that.

Would be even better for this to be broken out into methods, with a main method that shows the overall flow. But an overview description would the next best thing.

Reply:
I'll add some more comments about this in the version with broken-out methods, which I'll add to this branch soon: https://github.com/apache/druid/pull/12918#discussion_r951471931